File: 프로그래머스 level 3/합승 택시 요금.py
from collections import defaultdict

def solution(n, s, a, b, fares):
    fare_arr = make_map(n, fares)
    min_fare = float('Inf')
    for k in range(n):
        min_fare = min(fare_arr[s-1][k] + fare_arr[k][a-1] + fare_arr[k][b-1], min_fare)
    return min_fare


def make_map(n, fares):
    fare_arr = [[float('Inf')] * n for _ in range(n)]
    
    for fare in fares:
        fare_arr[fare[0]-1][fare[1]-1] = fare[-1]
        fare_arr[fare[1]-1][fare[0]-1] = fare[-1]
    
    for i in range(n):
        for j in range(n):
            if i == j:
                fare_arr[i][j] = 0
                
    for k in range(n):
        for i in range(n):
            for j in range(n):
                fare_arr[i][j] = min(fare_arr[i][k] + fare_arr[k][j], fare_arr[i][j])
    return fare_arr

File: Swift/프로그래머스 level3/합승택시요금.swift
import Foundation

func solution(_ n:Int, _ s:Int, _ a:Int, _ b:Int, _ fares:[[Int]]) -> Int {
    var distance = updateFare(n, fares)
    
    var minFare = Int.max
    
    for idx in 1..<n+1 {
        minFare = min(minFare, distance[s][idx] + distance[idx][a] + distance[idx][b])
    }
    
    return minFare
}

func updateFare(_ n: Int, _ fares: [[Int]]) -> [[Int]] {
    let inf = n * 100001 + 1
    
    var fareList = Array(repeating: Array(repeating: inf, count: n + 1), count: n + 1)
    
    for idx in 1..<n+1 {
        fareList[idx][idx] = 0
    }
    
    for fareInfo in fares {
        let (start, end, fare) = (fareInfo[0], fareInfo[1], fareInfo[2])
        (fareList[start][end], fareList[end][start]) = (fare, fare)
    }
    
    for k in 1..<n+1 {
        for i in 1..<n+1 {
            for j in 1..<n+1 {
                fareList[i][j] = min(fareList[i][j], fareList[i][k] + fareList[k][j])
            }
        }
    }
    
    return fareList
}